tempodb/backend/gcs/gcs.go

```diff
@@ -186,6 +190,112 @@ func (rw *readerWriter) List(ctx context.Context, keypath backend.KeyPath) ([]st
 	return objects, nil
 }
 
+// ListBlocks implements backend.Reader
+func (rw *readerWriter) ListBlocks(ctx context.Context, keypath backend.KeyPath) (blockIDs []uuid.UUID, compactedBlockIDs []uuid.UUID, err error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "readerWriter.ListBlocks")
+	defer span.Finish()
+
+	keypath = backend.KeyPathWithPrefix(keypath, rw.cfg.Prefix)
+	prefix := path.Join(keypath...)
+	if len(prefix) > 0 {
+		prefix += "/"
+	}
+
+	bb := blockboundary.CreateBlockBoundaries(rw.cfg.ListBlocksConcurrency)
+
+	errChan := make(chan error, len(bb))
+	wg := sync.WaitGroup{}
+	mtx := sync.Mutex{}
+
+	var min uuid.UUID
+	var max uuid.UUID
+
+	for i := 0; i < len(bb)-1; i++ {
+		min = uuid.UUID(bb[i])
+		max = uuid.UUID(bb[i+1])
+
+		wg.Add(1)
+		go func(min, max uuid.UUID) {
+			defer wg.Done()
+
+			iter := rw.bucket.Objects(ctx, &storage.Query{
+				Prefix:      prefix,
+				Delimiter:   "",
+				Versions:    false,
+				StartOffset: prefix + min.String(),
+			})
+
+			var parts []string
+			var id uuid.UUID
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					attrs, err := iter.Next()
+					if errors.Is(err, iterator.Done) {
+						return
+					}
+					if err != nil {
+						errChan <- fmt.Errorf("iterating blocks: %w", err)
+						return
+					}
+
+					parts = strings.Split(attrs.Name, "/")
+					// ie: <tenant>/<blockID>/meta.json
+					if len(parts) != 3 {
+						continue
+					}
+
+					if parts[2] != backend.MetaName && parts[2] != backend.CompactedMetaName {
+						continue
+					}
+
+					id, err = uuid.Parse(parts[1])
+					if err != nil {
+						continue
+					}
+
+					x := bytes.Compare(id[:], min[:])
+					if x < 0 {
+						return
+					}
+
+					y := bytes.Compare(id[:], max[:])
+					if y > 0 {
+						return
```

Comment:
if we've gone past the max block boundary should we exit the loop?

this comment applies to the s3 implementation as well.

Reply:
This return will stop the goroutine for the range.  We want the other routines to finish as well.

Reply:
> This return will stop the goroutine for the range. We want the other routines to finish as well.

if the guid is > max then this goroutine is done right? it should exit instead of churning through larger guids

Reply:
Yup. I see. You are correct :+1: